In [0]:
import numpy as np
import pandas as pd
import seaborn as sns

titanic = sns.load_dataset('titanic')

### 데이터 소스 : titanic_df = sns.load_dataset('titanic')
### Q. embark_town 열의 829행의 NaN 데이터의 앞, 뒤를 포함하는 3개행을 출력하세요.

In [0]:
print (titanic.embark_town[826:833])

826    Southampton
827      Cherbourg
828     Queenstown
829            NaN
830      Cherbourg
831    Southampton
832      Cherbourg
Name: embark_town, dtype: object


### Q. embark_town 열의 NaN값을 승선도시 중에서 가장 많이 출현한 값으로 대체하세요.

In [0]:
print (titanic.embark_town.isnull().sum())
print (titanic['embark_town'].value_counts().idxmax())
# print (titanic['embark_town'].median())
titanic.embark_town.replace(np.nan, value=titanic['embark_town'].value_counts().idxmax(), inplace=True)
print (titanic['embark_town'].value_counts())
print (titanic.embark_town.isnull().sum())

2
Southampton
Southampton    646
Cherbourg      168
Queenstown      77
Name: embark_town, dtype: int64
0


`replace` 대신 최빈값을 `drop` 한 뒤 `fillna`를 통해 최빈값으로 채울 수 있음

In [4]:
print (titanic.embark_town.value_counts().idxmin())

Queenstown


### Q. titanic_df에서 'age' 열에 나이 데이터가 없는 모든 행을 삭제 후 df_age 이름으로 저장하고 행의 갯수를 구하세요.  데이터 소스 : titanic_df = sns.load_dataset('titanic')

In [0]:
print (titanic['age'].unique())
print (titanic.age.count())
print ()
df_age = titanic.dropna(subset = ['age'], axis=0, inplace=False)
print (df_age.age.unique())
print (df_age.age.count())

[22.   38.   26.   35.     nan 54.    2.   27.   14.    4.   58.   20.
 39.   55.   31.   34.   15.   28.    8.   19.   40.   66.   42.   21.
 18.    3.    7.   49.   29.   65.   28.5   5.   11.   45.   17.   32.
 16.   25.    0.83 30.   33.   23.   24.   46.   59.   71.   37.   47.
 14.5  70.5  32.5  12.    9.   36.5  51.   55.5  40.5  44.    1.   61.
 56.   50.   36.   45.5  20.5  62.   41.   52.   63.   23.5   0.92 43.
 60.   10.   64.   13.   48.    0.75 53.   57.   80.   70.   24.5   6.
  0.67 30.5   0.42 34.5  74.  ]
714

[22.   38.   26.   35.   54.    2.   27.   14.    4.   58.   20.   39.
 55.   31.   34.   15.   28.    8.   19.   40.   66.   42.   21.   18.
  3.    7.   49.   29.   65.   28.5   5.   11.   45.   17.   32.   16.
 25.    0.83 30.   33.   23.   24.   46.   59.   71.   37.   47.   14.5
 70.5  32.5  12.    9.   36.5  51.   55.5  40.5  44.    1.   61.   56.
 50.   36.   45.5  20.5  62.   41.   52.   63.   23.5   0.92 43.   60.
 10.   64.   13.   48.    0.75 53.   57

### Q. age 열의 NaN값을 다른 나이 데이터의 평균으로 변경한 후 'age' 열의 첫 10개 데이터만 정수로 출력하세요.

In [0]:
titanic.age.replace(np.nan, value = titanic.age.mean(), inplace=True)
print (titanic.age.head(10).astype(int))

0    22
1    38
2    26
3    35
4    35
5    35
6    54
7     2
8    27
9    14
Name: age, dtype: int64


```
titanic.age.fillna(titanic.age.mean(), inplace=True)
```

### Q. 1 ~ 2사이의 정수로 (5,3) 2차원 배열을 랜덤하게 만든 후 DataFrame으로 불러와 df에 저장하고 df에서 중복 행을 제거한 후 df2에 저장 후 출력하세요.
 - columns = ['a','b','c']

In [6]:
np.random.seed(0)
df = pd.DataFrame(np.random.randint(1,3, size=(5,3)), columns=['a','b','c'])
print (df, '\n', type(df), '\n')

df2 = df.drop_duplicates()
print(df2)

   a  b  c
0  1  2  2
1  1  2  2
2  2  2  2
3  2  2  1
4  1  2  1 
 <class 'pandas.core.frame.DataFrame'> 

   a  b  c
0  1  2  2
2  2  2  2
3  2  2  1
4  1  2  1


### Q. df에서 b, c열을 기준으로 중복 행을 제거한 후 df3에 저장하고 출력하세요.

In [9]:
df3 = df.drop_duplicates(subset = ['b','c'])
print (df3, '\n')

print (df.drop_duplicates(subset = ['a','b']), '\n')

print (df.drop_duplicates(subset = ['b']))

   a  b  c
0  1  2  2
3  2  2  1 

   a  b  c
0  1  2  2
2  2  2  2 

   a  b  c
0  1  2  2


### Q. titanic_df embark_town 열의 NaN값을 바로 앞에 있는 830행의 값으로 변경한 후 출력하세요.

In [0]:
titanic = sns.load_dataset('titanic')
print (titanic.embark_town[826:833],'\n')
titanic.embark_town.fillna(method = 'ffill', inplace=True)
print (titanic.embark_town[826:833])

826    Southampton
827      Cherbourg
828     Queenstown
829            NaN
830      Cherbourg
831    Southampton
832      Cherbourg
Name: embark_town, dtype: object 

826    Southampton
827      Cherbourg
828     Queenstown
829     Queenstown
830      Cherbourg
831    Southampton
832      Cherbourg
Name: embark_town, dtype: object


### Q. horsepower 열의 누락 데이터 '?'을 삭제한 후 NaN 값의 갯수를 출력하세요.

In [10]:
mpg = sns.load_dataset('mpg')
mpg.columns = ['mpg','cylinders','displacement','horsepower','weight',
              'acceleration','model year','origin','name'] 
print (mpg.horsepower.unique(), '\n')
print(mpg.horsepower.isnull().sum(), '\n')

mpg.dropna(subset=['horsepower'], axis=0 , inplace=True)
print (mpg.horsepower.unique(), '\n')
print(mpg.horsepower.isnull().sum(), '\n')
print (mpg.horsepower.value_counts(dropna=False))

[130. 165. 150. 140. 198. 220. 215. 225. 190. 170. 160.  95.  97.  85.
  88.  46.  87.  90. 113. 200. 210. 193.  nan 100. 105. 175. 153. 180.
 110.  72.  86.  70.  76.  65.  69.  60.  80.  54. 208. 155. 112.  92.
 145. 137. 158. 167.  94. 107. 230.  49.  75.  91. 122.  67.  83.  78.
  52.  61.  93. 148. 129.  96.  71.  98. 115.  53.  81.  79. 120. 152.
 102. 108.  68.  58. 149.  89.  63.  48.  66. 139. 103. 125. 133. 138.
 135. 142.  77.  62. 132.  84.  64.  74. 116.  82.] 

6 

[130. 165. 150. 140. 198. 220. 215. 225. 190. 170. 160.  95.  97.  85.
  88.  46.  87.  90. 113. 200. 210. 193. 100. 105. 175. 153. 180. 110.
  72.  86.  70.  76.  65.  69.  60.  80.  54. 208. 155. 112.  92. 145.
 137. 158. 167.  94. 107. 230.  49.  75.  91. 122.  67.  83.  78.  52.
  61.  93. 148. 129.  96.  71.  98. 115.  53.  81.  79. 120. 152. 102.
 108.  68.  58. 149.  89.  63.  48.  66. 139. 103. 125. 133. 138. 135.
 142.  77.  62. 132.  84.  64.  74. 116.  82.] 

0 

150.0    22
90.0     20
88.0     19
1

### Q. horsepower 열의 누락 데이터('?') 삭제하고 실수형으로 변환하세요.

In [0]:
mpg = sns.load_dataset('mpg')
# print (type(mpg))
mpg.columns = ['mpg','cylinders','displacement','horsepower','weight',
              'acceleration','model year','origin','name'] 
# print (mpg.unique(), '\n')
print (mpg.horsepower.unique(), '\n')

mpg.dropna(subset=['horsepower'], axis=0, inplace=True)
print (mpg.horsepower.unique(), '\n')
mpg.horsepower.astype(float)
print (mpg.horsepower.dtype)

[130. 165. 150. 140. 198. 220. 215. 225. 190. 170. 160.  95.  97.  85.
  88.  46.  87.  90. 113. 200. 210. 193.  nan 100. 105. 175. 153. 180.
 110.  72.  86.  70.  76.  65.  69.  60.  80.  54. 208. 155. 112.  92.
 145. 137. 158. 167.  94. 107. 230.  49.  75.  91. 122.  67.  83.  78.
  52.  61.  93. 148. 129.  96.  71.  98. 115.  53.  81.  79. 120. 152.
 102. 108.  68.  58. 149.  89.  63.  48.  66. 139. 103. 125. 133. 138.
 135. 142.  77.  62. 132.  84.  64.  74. 116.  82.] 

[130. 165. 150. 140. 198. 220. 215. 225. 190. 170. 160.  95.  97.  85.
  88.  46.  87.  90. 113. 200. 210. 193. 100. 105. 175. 153. 180. 110.
  72.  86.  70.  76.  65.  69.  60.  80.  54. 208. 155. 112.  92. 145.
 137. 158. 167.  94. 107. 230.  49.  75.  91. 122.  67.  83.  78.  52.
  61.  93. 148. 129.  96.  71.  98. 115.  53.  81.  79. 120. 152. 102.
 108.  68.  58. 149.  89.  63.  48.  66. 139. 103. 125. 133. 138. 135.
 142.  77.  62. 132.  84.  64.  74. 116.  82.] 

float64


### Q. Titanic 가공한 내역을 정리하고 이를 쉽게 재사용할 후 있도록 종합 전처리 함수를 만드세요.(def transform_features(df):)

In [0]:
titanic = sns.load_dataset('titanic')
print (titanic.info())
print (titanic.sample(10))

In [18]:
%%time

from sklearn.preprocessing import LabelEncoder

titanic = sns.load_dataset('titanic')
# print (titanic.columns.tolist())
# print (titanic.describe())

def fillna(df) :
    df['age'].fillna(df.age.mean(), axis = 0, inplace=True)
    # print (df.age.unique())
    df['embark_town'].fillna(method='ffill', axis = 0, inplace=True)
    # print (df['embark_town'].unique())
    return df

def drop_features(df) :
    df.drop(['embarked', 'class', 'deck','alone','who', 'alive','adult_male'], axis=1, inplace=True)
    return df

def format_features(df) :
    formats = df.columns.tolist()
    # print (formats)
    for f in formats :
        # print(df[f].dtype)
        if df[f].dtype == 'object' : 
            df[f].astype('category')
            # print (df[f], '\n', type(df[f]))
            df[f] = LabelEncoder().fit_transform(df[f])
    return df

def transform_features(df):
    df = format_features(drop_features(fillna(df)))
    return df

titanic = transform_features(titanic)
print (titanic.sample(5),'\n')
print (titanic.info(), '\n')

     survived  pclass  sex        age  sibsp  parch     fare  embark_town
251         0       3    0  29.000000      1      1  10.4625            2
460         1       1    1  48.000000      0      0  26.5500            2
496         1       1    0  54.000000      1      0  78.2667            0
485         0       3    0  29.699118      3      1  25.4667            2
144         0       2    1  18.000000      0      0  11.5000            2 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 8 columns):
survived       891 non-null int64
pclass         891 non-null int64
sex            891 non-null int64
age            891 non-null float64
sibsp          891 non-null int64
parch          891 non-null int64
fare           891 non-null float64
embark_town    891 non-null int64
dtypes: float64(2), int64(6)
memory usage: 55.8 KB
None 

CPU times: user 37.4 ms, sys: 3.97 ms, total: 41.4 ms
Wall time: 107 ms


## ##

In [0]:
from google.colab import drive
drive.mount('/content/drive')

In [46]:
train = pd.read_csv('/content/drive/My Drive/Colab Notebooks/python-lecture/HW/train.csv')
train.drop(['PassengerId', 'Name', 'Ticket'], axis = 1, inplace=True)
print (train.columns)
print (train.columns.values)
print (type(train.columns.values))

Index(['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin',
       'Embarked'],
      dtype='object')
['Survived' 'Pclass' 'Sex' 'Age' 'SibSp' 'Parch' 'Fare' 'Cabin' 'Embarked']
<class 'numpy.ndarray'>


In [47]:
print (train.Age, '\n')
print (train.Age.isnull().sum(), '\n')
print (train.Age.mean(), '\n')
train.Age.fillna(train.Age.mean(), inplace=True)
print (train.Age.isnull().sum(), '\n')

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64 

177 

29.69911764705882 

0 



In [48]:
print (train.Cabin, '\n')
print (train.Cabin.isnull().sum(), '\n')
train.Cabin.fillna('N', inplace=True)
print (train.Cabin.isnull().sum(), '\n')
print (train.Cabin.value_counts(dropna=False), '\n')

0       NaN
1       C85
2       NaN
3      C123
4       NaN
       ... 
886     NaN
887     B42
888     NaN
889    C148
890     NaN
Name: Cabin, Length: 891, dtype: object 

687 

0 

N              687
G6               4
C23 C25 C27      4
B96 B98          4
F33              3
              ... 
D46              1
C104             1
C111             1
D30              1
T                1
Name: Cabin, Length: 148, dtype: int64 



In [49]:
print (train.Embarked, '\n')
print (train.Embarked.isnull().sum(), '\n')
train.Embarked.fillna(method='ffill', inplace=True)
print (train.Embarked.isnull().sum(), '\n')
print (train.Embarked.value_counts(), '\n')

0      S
1      C
2      S
3      S
4      S
      ..
886    S
887    S
888    S
889    C
890    Q
Name: Embarked, Length: 891, dtype: object 

2 

0 

S    644
C    169
Q     78
Name: Embarked, dtype: int64 



In [50]:
from sklearn.preprocessing import LabelEncoder

# lft = LabelEncoder().fit_transform

train.Cabin = LabelEncoder().fit_transform( train.Cabin.str[:1] )
print (train.Cabin, '\n')

train.Sex = LabelEncoder().fit_transform( train.Sex )
print (train.Sex, '\n')

train.Embarked = LabelEncoder().fit_transform(train.Embarked)
print (train.Embarked, '\n')

print (train.sample(10), '\n')

0      7
1      2
2      7
3      2
4      7
      ..
886    7
887    1
888    7
889    2
890    7
Name: Cabin, Length: 891, dtype: int64 

0      1
1      0
2      0
3      0
4      1
      ..
886    1
887    0
888    0
889    1
890    1
Name: Sex, Length: 891, dtype: int64 

0      2
1      0
2      2
3      2
4      2
      ..
886    2
887    2
888    2
889    0
890    1
Name: Embarked, Length: 891, dtype: int64 

     Survived  Pclass  Sex        Age  SibSp  Parch     Fare  Cabin  Embarked
467         0       1    1  56.000000      0      0  26.5500      7         2
705         0       2    1  39.000000      0      0  26.0000      7         2
45          0       3    1  29.699118      0      0   8.0500      7         2
18          0       3    0  31.000000      1      0  18.0000      7         2
339         0       1    1  45.000000      0      0  35.5000      8         2
500         0       3    1  17.000000      0      0   8.6625      7         2
209         1       1    1  40.00

## Q. titanic의 각 컬럼별 null 개수를 for loop 사용해서 구한 후 출력

In [17]:
titanic = sns.load_dataset('titanic')

missings = titanic.isnull()
# print (missings)
for col in missings.columns :
    mc = missings[col].value_counts()
    try : print (col, '\t: ', mc[True])
    except : print (col, '\t: ', 0)

survived 	:  0
pclass 	:  0
sex 	:  0
age 	:  177
sibsp 	:  0
parch 	:  0
fare 	:  0
embarked 	:  2
class 	:  0
who 	:  0
adult_male 	:  0
deck 	:  688
embark_town 	:  2
alive 	:  0
alone 	:  0


In [16]:
print (titanic.isnull().sum())

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64
